---
title: "KEGG analysis of Arabidopsis RNAseq data"
author: "M. Naoumkina"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    number_sections: true
    code_download: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

The count data of RNAseq was downloaded from [Expression Atlas web page](https://www.ebi.ac.uk/gxa/experiments/E-GEOD-72806/Downloads)

## Load required packages

```{r, message=FALSE, warning=FALSE}
#BiocManager::install("clusterProfiler") 
#BiocManager::install("pathview") 
#BiocManager::install("enrichplot")

library(pathview)
library(clusterProfiler)
library(enrichplot)
library(ggplot2)
```

## Set the desired organism

```{r, message=FALSE, warning=FALSE}
library(org.At.tair.db)
organism <- "org.At.tair.db"
if (!requireNamespace(organism, quietly = TRUE)) {
  BiocManager::install(organism, character.only = TRUE)}
library(organism, character.only = TRUE)
```

## Prepare Input

```{r, message=FALSE, warning=FALSE}
# Load saved data DE_res_heat, column name X as gene IDs
df = read.csv("DE_res_heat.csv", header=TRUE)
# we want the log2 fold change 
original_gene_list <- df$logFC
# name the vector
names(original_gene_list) <- df$X
# omit any NA values 
gene_list<-na.omit(original_gene_list)
# check the gene_list is named numeric vector
str(gene_list)
#Filter out non-nuclear genes
gene_list <- gene_list[grep("^AT[1-5]G", names(gene_list))]
# sort the list in decreasing order (required for clusterProfiler)
gene_list = sort(gene_list, decreasing = TRUE)
```
## Convert gene IDs for gseKEGG function

```{r, message=FALSE, warning=FALSE}
# We will lose some genes here because not all IDs will be converted
ids<-bitr(names(gene_list), fromType = "TAIR", toType = "ENTREZID", OrgDb=organism)
# remove duplicate IDS (here I use "TAIR", but it should be whatever was selected as keyType)
dedup_ids = ids[!duplicated(ids[c("TAIR")]),]

# Create a new dataframe df2 which has only the genes which were successfully mapped using the bitr function above
df2 = df[df$X %in% dedup_ids$TAIR,]

# Create a new column (column Y = ENTREZ) in df2 with the corresponding ENTREZ IDs
df2$Y = dedup_ids$ENTREZID
```

## Create a vector of the kegg gene list
```{r, message=FALSE, warning=FALSE}
kegg_gene_list <- df2$logFC
# Name vector with ENTREZ ids
names(kegg_gene_list) <- df2$Y
# omit any NA values 
kegg_gene_list<-na.omit(kegg_gene_list)
# remove duplicates
kegg_gene_list <- kegg_gene_list[!duplicated(names(kegg_gene_list))]
head(kegg_gene_list)
any(duplicated(names(kegg_gene_list)))  # should return FALSE
# sort the list in decreasing order (required for clusterProfiler)
kegg_gene_list <- sort(kegg_gene_list, decreasing = TRUE)
```
## Create gseKEGG object
KEGG Organism Code. For Arabidopsis the code is 'ath'. 
The full list of organisms codes is [here](https://www.genome.jp/kegg/catalog/org_list.html). 
 - I define this as kegg_organism first, because it is used again below when making the pathview plots.
 - nPerm the higher the number of permutations you set, the more accurate your result will, but the longer the analysis will take.
 - minGSSize minimum number of genes in set (gene sets with lower than this many genes in your dataset will be ignored).
 - maxGSSize maximum number of genes in set (gene sets with greater than this many genes in your dataset will be ignored).
 - pvalueCutoff pvalue Cutoff.
 - pAdjustMethod one of “holm”, “hochberg”, “hommel”, “bonferroni”, “BH”, “BY”, “fdr”, “none”.
 - keyType one of ‘kegg’, ‘ncbi-geneid’, ‘ncib-proteinid’ or ‘uniprot’.

```{r,message=FALSE, warning=FALSE}
kegg_organism = "ath"
kk2 <- gseKEGG(geneList     = kegg_gene_list,
               organism     = kegg_organism,
               nPerm        = 10000,
               minGSSize    = 3,
               maxGSSize    = 800,
               pvalueCutoff = 0.05,
               pAdjustMethod = "none",
               keyType       = "ncbi-geneid")

#To evaluate kk2 object
head(as.data.frame(kk2), 10)
```

## Dotplot with forced one-line pathway labels
```{r, fig.width=12, message=FALSE, warning=FALSE}
library(stringr)  # for str_wrap
dotplot(kk2, showCategory = 10, title = "Enriched Pathways" , split=".sign") + facet_grid(.~.sign) + theme(
  axis.text.y = element_text(size = 10),
  plot.margin = margin(5, 5, 5, 20, "pt")
) +
  scale_y_discrete(labels = function(x) str_wrap(x, width = 400))
```

## Encrichment map
```{r, message=FALSE, warning=FALSE}
# check number of enrichment terms
nrow(kk2@result)
# Compute Term Similarity Explicitly
kk2 <- pairwise_termsim(kk2)
# run emapplot 
emapplot(kk2, showCategory = 20)
```

## Category Netplot
```{r, message=FALSE, warning=FALSE}
# categorySize can be either 'pvalue' or 'geneNum'
cnetplot(kk2, categorySize="pvalue", foldChange=gene_list)
```

## Ridgeplot with forced one-line pathway labels
```{r, message=FALSE,warning=FALSE}
library(stringr)  # for str_wrap
ridgeplot(kk2) +
  labs(x = "Enrichment distribution") +
  theme(axis.text.y = element_text(size = 10)) +
  scale_y_discrete(labels = function(x) str_wrap(x, width = 400))
```

## GSEA Plot
Gene Set Integer corresponds to gene set in the GSE object. 
The first gene set is 1, second gene set is 2, etc. Default: 1
Use the `Gene Set` param for the index in the title, and as the value for `geneSetId`.

```{r, fig.width=10, fig.height=6, message=FALSE, warning=FALSE}
gseaplot(kk2, by = "all", title = kk2$Description[1], geneSetID = 1)
```

## Pathview
Following parameters are needed to be enter: `gene.data` is the kegg_gene_list created above; `pathway.id`, which cab be found in the gseKEGG output table; `species` same as organism above in gseKEGG, which we defined as kegg_organism.

```{r, echo=TRUE, out.width="100%", warning=FALSE, message=FALSE}
# Check pathway IDs first
kk2@result$ID

pathview(gene.data = kegg_gene_list,
         pathway.id = "ath00750",
         species = "ath",
         kegg.native = FALSE)

knitr::include_graphics("ath00750.pathview.png")
```

```{r, message=FALSE, warning=FALSE}
# Save the native KEGG plot (PNG) in your working directory
ath <- pathview(gene.data=kegg_gene_list, pathway.id="ath00750", species = kegg_organism)
```

## Session Info
```{r session-info, echo=TRUE}
sessionInfo()
```
